Chepa kn unit2

ChepaKN opened the pull request:
Unit2 pull request

ChepaKN committed:
unit1. reverse() realization.

ChepaKN committed:
Домашнее задание №2.
1.Удалил метод hit(). Вместо этого цель(Target) принимает(receiveDamage) контейнер(Damage), содержащий структурированный урон и название оружия, которым он нанесен, там же учитывается индивидуальный иммунитет цели.
2.Для физического урона реализовал случайную составляющую и возможность нанесения критического удара.
3.Добавил несколько тестов.

Bael commented on unit1/src/main/java/com/github/siberianintegrationsystems/Unit1.java:
```diff
@@ -15,7 +15,12 @@ public static void main(String[] args) {
-        return "";
+        int len = INPUT_STRING.length();
+        char[] buf = new char[len];
+        for(int c = 0; c < len; c++){
```
индексы лучше именовать i, j, k, l и т.д. другим людям читающий ваш код будет проще

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Damage.java:
```diff
@@ -0,0 +1,45 @@
+
+    public Damage(String armName, long phys, long fire, long ice){
+        this.armName = armName;
+        this.phys   = phys;
```
не экономьте символы, слова damage сильно не хватает.

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Main.java:
```diff
@@ -0,0 +1,28 @@
+
+    public static void main(String[] args) {
+
+        // Сценарий битвы
```
за сценарий жирный плюс :)

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Target.java:
```diff
@@ -1,22 +1,80 @@
 
-    public int getPhysicalDamage() {
-        return physicalDamage;
+    private boolean         alive;
```
форматирование красивое, но слетит при автоформатировании в идее, поэтому не советую тратить время на это.

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Target.java:
```diff
@@ -1,22 +1,80 @@
+        this.healthPoint    = healthPoint;
+        this.immunity       = immunity;
+
+        if(healthPoint > 0) this.alive = true;
```
не жалейте скобок :)

Bael replied:
тут по смыслу проще и короче написать this.alive = healthPoint > 0;

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Target.java:
```diff
@@ -1,22 +1,80 @@
+        //Урон с учетом имунитета
+        long totalDamage = 0;
+        switch (immunity){
+            case("fire"):
```
в строках люди чаще ошибаются, лучше использовать енамы.

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Target.java:
```diff
@@ -1,22 +1,80 @@
+
+        healthPoint -= totalDamage;
+        if(healthPoint <= 0) {
+            System.out.println(this.name + ":"  + "\tУбит оружием: \t" + damage.getArmName());
```
хорошая идея, фиксировать убийцу.

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Weapon.java:
```diff
@@ -1,7 +1,65 @@
+    }
+
+    public Weapon() {
+        this.name   = "Безобидный перочинынй нож";
```
перочинный 
ошибки в коде живут годами, лучше проверять сразу, чтобы потом стыдно не было

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Weapon.java:
```diff
@@ -1,7 +1,65 @@
+    public Weapon() {
+        this.name   = "Безобидный перочинынй нож";
+        this.damage = new Damage("", 0,0,0);
+        readyToWar = false;
```
Идея интересная, но по сути лишняя. Достаточно чтобы наносимый урон был равен нулю.
Насчет безобидности, в морровинде был квест, где нужно было ложкой убить летающего монстра.

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Weapon.java:
```diff
@@ -1,7 +1,65 @@
+
+    //Дополнительный "Критический" физический урон
+    private long getCriticalDamage(){
+        long random1 = (long)(Math.random()*20);
```
проще работать на int, там достаточно разрядов для любого урона до нескольких миллиардов насколько я помню.

Bael replied:
я бы тут умножал основной урон на некий коээфициент. тогда безобидный перочинный ножик тоже бы ничего не мог сделать

Bael commented on unit2/homework/src/main/java/pro/it/sis/javacourse/Main.java:
```diff
@@ -0,0 +1,28 @@
+                while (t.isAlive()) {
+                    System.out.println(t.toString());
+                    t.receiveDamage(swords.get(0).hurt());
+                }t.raise(1000);
```
надо поправить форматирование

Bael reviewed (approved):
В целом ваш подход будет работать, даже без наследования.
Но тогда вам нужен класс который будет создавать экземпляры оружия и цели с нужными параметрами - фабрика.
Вы же не будете просить клиента вашего кода запомнить характеристики всего оружия в системе?

Из остальных замечаний - замечания к форматированию, видно что вы пишете на чем то другом, и это немного мешает.

Хочу похвалить за продуманную составляющую игровой механики. Творческий режим у вас сработал на 200% :)

ChepaKN committed:
1. Добавил фабрики, создающие объекты оружия и целей.
2. Внес поправки согласно замечаниям.
Считаю, что наследование в данном примере избыточно. Другое дело, если бы создавать объект воина (цель + оружие). Но как я понимаю, множественное наследование в java недоступно... Вобщем можно пофантазировать на досуге:)
